File: app/src/main/java/com/goel/attendancetracker/downloadmanager/ClassDownloadManager.java
package com.goel.attendancetracker.downloadmanager;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.pdf.PdfDocument;
import android.os.Environment;

import com.goel.attendancetracker.classes.ClassesModel;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class ClassDownloadManager extends FileDataModel {
    public static final int WIDTH = 740;
    public static final int HEIGHT = 1300;
    public static final int TITLE_PAGE_HEIGHT = 550;
    public static final int DOWNLOAD_SUCCESSFUL = 1;
    public static final int DOWNLOAD_FAILED = -1;
    private String filePath;
    private final ClassesModel model;
    private PdfDocument document;
    private Paint paint;
    private PdfDocument.PageInfo pageInfo;
    private PdfDocument.Page page;
    private int pageNumber;
    private Canvas canvas;
    private final String organisationName;

    public ClassDownloadManager(ClassesModel model, String organisationName) {
        this.model = model;
        this.organisationName = organisationName;
        this.pageNumber = 0;
        this.setFilePath();
    }

    private void setFilePath(){
        this.filePath = "Attendance Tracker" + "/" + this.model.getClassName() + ".pdf";
    }

    public String getFilePath(){
        return this.filePath;
    }

    public int downloadAttendance(){
        this.initializeDocument();
        File file = new File(Environment.getExternalStorageDirectory(), "/" + this.getFilePath());
        this.addPage();
        this.writeToPage();
        try {
            downloadFile(file);
            this.document.close();
            return DOWNLOAD_SUCCESSFUL;
        }
        catch (IOException e) {
            e.printStackTrace();
            this.document.close();
            return DOWNLOAD_FAILED;
        }
    }

    private void initializeDocument(){
        this.document = new PdfDocument();
        this.paint = new Paint();
    }

    public void addPage()  {
        this.pageNumber++;
        int height = this.pageNumber == 1 ? TITLE_PAGE_HEIGHT : HEIGHT;
        this.pageInfo = new PdfDocument.PageInfo.Builder(WIDTH, height, this.pageNumber).create();
        this.page = document.startPage(pageInfo);
        this.canvas = page.getCanvas();
    }

    private void writeToPage(){
        this.createFirstPage(this);
        this.document.finishPage(this.page);
        this.writeAttendance(this);
    }

    private void downloadFile(File file) throws IOException {
        try {
            this.document.writeTo(new FileOutputStream(file));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            (new File(Environment.getExternalStorageDirectory(), "/Attendance Tracker")).mkdirs();
            this.document.writeTo(new FileOutputStream(file));
        }
    }

    // ===================== GETTERS AND SETTERS ===================== //

    public ClassesModel getModel() {
        return model;
    }

    public String getOrganisationName() {
        return organisationName;
    }

    public PdfDocument getDocument() {
        return document;
    }

    public Paint getPaint() {
        return paint;
    }

    public PdfDocument.PageInfo getPageInfo() {
        return pageInfo;
    }

    public PdfDocument.Page getPage() {
        return page;
    }

    public Canvas getCanvas() {
        return canvas;
    }
}

File: app/src/main/java/com/goel/attendancetracker/downloadmanager/ClassDownloadManager.kt
package com.goel.attendancetracker.downloadmanager

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.pdf.PdfDocument
import android.graphics.pdf.PdfDocument.PageInfo
import android.os.Environment
import com.goel.attendancetracker.models.ClassesModel
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException

class ClassDownloadManager(// ===================== GETTERS AND SETTERS ===================== //
    val model: ClassesModel, val organisationName: String
) : FileDataModel() {
    val filePath = "Attendance Tracker/$organisationName/${model.className}.pdf"
    var document: PdfDocument? = null
        private set
    lateinit var paint: Paint
        private set
    lateinit var pageInfo: PageInfo
        private set
    lateinit var page: PdfDocument.Page
        private set
    private var pageNumber = 0
    lateinit var canvas: Canvas
        private set

    fun downloadAttendance(): Int {
        initializeDocument()
        val file = File(Environment.getExternalStorageDirectory(), "/$filePath")
        addPage(false)
        writeToPage()
        return try {
            downloadFile(file)
            document!!.close()
            DOWNLOAD_SUCCESSFUL
        } catch (e: IOException) {
            e.printStackTrace()
            document!!.close()
            DOWNLOAD_FAILED
        }
    }

    private fun initializeDocument() {
        document = PdfDocument()
        paint = Paint()
    }

    fun addPage(isLastPage: Boolean) {
        pageNumber++
        val height: Int = when {
            pageNumber == 1 -> {
                TITLE_PAGE_HEIGHT
            }
            isLastPage -> {
                Last_PAGE_HEIGHT
            }
            else -> {
                HEIGHT
            }
        }
        pageInfo = PageInfo.Builder(WIDTH, height, pageNumber).create()
        page = document!!.startPage(pageInfo)
        canvas = page.canvas
    }

    private fun writeToPage() {
        createFirstPage(this)
        document!!.finishPage(page)
        writeAttendance(this)
    }

    @Throws(IOException::class)
    private fun downloadFile(file: File) {
        try {
            document!!.writeTo(FileOutputStream(file))
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
            File(
                Environment.getExternalStorageDirectory(),
                "/Attendance Tracker/$organisationName"
            )
                .mkdirs()
            document!!.writeTo(FileOutputStream(file))
        }
    }

    companion object {
        private const val WIDTH = 740
        private const val HEIGHT = 1300
        private const val TITLE_PAGE_HEIGHT = 550
        private const val Last_PAGE_HEIGHT = 300
        const val DOWNLOAD_SUCCESSFUL = 1
        const val DOWNLOAD_FAILED = -1
    }
}